### ISSUE TITLE
SubJet issues

### ISSUE THREAD

**Issue by intrepid42** [cms-nanoAOD/cmssw#194]
Hi, I see some strange issues with the SubJet collection.

There are some subJet_Idx larger than nSubJet (-> out of range):
![image](https://user-images.githubusercontent.com/1311783/44709499-12b7d580-aaaa-11e8-814a-9001c17333f7.png)
![image](https://user-images.githubusercontent.com/1311783/44709510-1b101080-aaaa-11e8-820b-9e2abde79402.png)

There are SubJets of the same FatJet that have large DeltaR (and this population is even visible on linear scale when we select candidates for boosted top/W):
![image](https://user-images.githubusercontent.com/1311783/44709756-ac7f8280-aaaa-11e8-8109-037694a4cb50.png)
`Events->Draw("sqrt(pow(TVector2::Phi_mpi_pi(SubJet_phi[FatJet_subJetIdx1]-SubJet_phi[FatJet_subJetIdx2]),2)+pow(SubJet_eta[FatJet_subJetIdx1]-SubJet_eta[FatJet_subJetIdx2],2))", "FatJet_subJetIdx1>=0 && FatJet_subJetIdx2>=0 && FatJet_subJetIdx1<nSubJet && FatJet_subJetIdx2<nSubJet")`

Maybe there are cuts applied that remove (low-pt?) subjets and break the indices?

@rappoccio @clelange

edit: I used this input file: `/store/mc/RunIISummer16NanoAOD/TT_TuneCUETP8M2T4_13TeV-powheg-pythia8/NANOAODSIM/PUMoriond17_05Feb2018_94X_mcRun2_asymptotic_v2-v1/80000/F63FB165-6D10-E811-907E-FA163E247A2E.root`

**Comment by rappoccio** [comment]
Hmmm... Well, they're clearly not part of the same jet (dr < 0.8 for real subjets). 

Can you try adding a pt > 170 GeV cut to the Fat Jets to see if that improves?

**Comment by intrepid42** [comment]
Hi Sal, that leaves the plot unchanged (as `pt > 170` for all FatJets).

**Comment by rappoccio** [comment]
Not all FatJets, the leading three are always kept, regardless of pt. Hmm. I need to investigate further.

**Comment by rappoccio** [comment]
Also, BTW, this has to be a miniaod issue, nanoaod is just "forwarding" the information... but I will take a look.

**Comment by intrepid42** [comment]
That's scary... Thanks for checking!

**Comment by rappoccio** [comment]
OK, good news first: This is not present in miniaod. 

Still digging into what happened in NANOAOD.

**Comment by rappoccio** [comment]
I figured this out. 

In the past, we had two subjet collections (CHS, and PUPPI). See [here](https://github.com/cms-sw/cmssw/commit/903330a6247fd886c8ffca0f1fafc0b11bc74d05#diff-416ad16d53535660648a4a4bb98636d6L240). We switched to just one ```'SoftDropPuppi'``` for later releases. Explicitly asking for this fixes the problem. 

So this will occur if somehow PUPPI removes a subjet that CHS kept, and the indices get out of sync. It occurs very rarely, but does occur. If desired it is still okay to do a delta-R matching between the subjets and the ungroomed fat jets as a workaround. 

Sorry about that, my mistake. This occurred so rarely I didn't catch it. It is probably sufficient for now to either use the delta-R matching, or just checking that the subjets are within 0.8 of the fat jet itself as a sanity check.
